content/blog/foreign-key-constraints.md

```diff
@@ -0,0 +1,198 @@
+ table_with_two_unique_first_name_key         | u               | table_with_two_unique | first_name        |                  |                   | UNIQUE (first_name)
+
+```
+
+Note in particular that `constraint_table` and `constraint_column` are identical for both the indvidual unique key constraints and the multi-column unique key constraint.
+
+
+## SQLite
+
+SQLite does not have catalog tables like postgres so we need a slightly different appraoch for extracting foreign key constraints. We can use `PRAGMA` statements instead to query the database metadata. The `PRAGMA` statement we will use is `PRAGMA foreign_key_list('table_name')` which retrieves foreign key information for a single table. Unfortunately there isn't a way to get foreign keys constraints for all tables at once so we need to do some work up front to get the list of tables and then loop over this to retrieve the foreign key constraints for each of the tables in turn.
+
+The `PRAGMA` statement can be called using `PRAGMA foreign_key_list('table_name')` or if using `sqlite3` `SELECT * FROM pragma_foreign_key_list('table_name')`. We will use the second form as this allows us to `SELECT` the specific columns we're interested in.
+
+Firstly we need to know our list of tables. This can be retrieved from the `sqlite_master` table using `SELECT tbl_name FROM sqlite_master WHERE type = 'table';`.
+
+For each returned table we now need to use the `PRAGMA` to get the foreign keys:
+```
+SELECT 
+  'table_name' AS constraint_table,
+  pragma.'from' AS constraint_column,
+  pragma.'table' AS referenced_table,
+  pragma.'to' AS referenced_column
+FROM  pragma_foreign_key_list('table_name') AS pragma;
+```
+
+Substituting `table_name` for the name of each table in turn. Note that the quotes around `table`, `from` and `to` are required as these are protected words in SQLite so need to be quoted to be interpreted as a string literal.
+
+Using the same setup code as above to create the three test tables we can use the query for each table and `UNION` them to get:
+
+```
+SELECT 
+  'region' AS constraint_table,
+  pragma.'from' AS constraint_column,
+  pragma.'table' AS referenced_table,
+  pragma.'to' AS referenced_column
+FROM  pragma_foreign_key_list('region') AS pragma
+
+UNION
+
+SELECT 
+  'street' AS constraint_table,
+  pragma.'from' AS constraint_column,
+  pragma.'table' AS referenced_table,
+  pragma.'to' AS referenced_column
+FROM  pragma_foreign_key_list('street') AS pragma
+
+UNION
+
+SELECT 
+  'address' AS constraint_table,
+  pragma.'from' AS constraint_column,
+  pragma.'table' AS referenced_table,
+  pragma.'to' AS referenced_column
+FROM  pragma_foreign_key_list('address') AS pragma;
+
+```
+
+Running with `.headers ON` returns:
+```
+constraint_table|constraint_column|referenced_table|referenced_column
+address|region|region|name
+address|street|street|name
+region|parent|region|name
+```
```

Comment:
This sort of begs for some sort of conclusion - returning back to what we were trying to achieve through all the detail